exercises/say/.meta/src/reference/java/Say.java

```diff
@@ -0,0 +1,204 @@
+            char headDigit = digits[0];
+            char[] tailDigits = new char[2];
+            tailDigits[0] = digits[1];
+            tailDigits[1] = digits[2];
+            digits = filterInitialZeros(tailDigits);
+            sentence = convertUnit(headDigit) + " hundred";
+        }
+
+        boolean oneDigit = digits.length == 1;
+        boolean twoDigits = digits.length == 2;
+
+        if (twoDigits) {
+            sentence = concatWithSpace(sentence, sayTwoDigits(digits));
+        }
+        if (oneDigit) {
+            sentence = concatWithSpace(sentence, convertUnit(digits[0]));
+        }
+
+        if (sentence.length() != 0) {
+            return concatWithSpace(sentence, order(orderNumber));
+        } else {
+            return sentence;
+        }
+    }
+
+    private String sayTwoDigits(char[] digits) {
+        if (digits[0] == '1') {
+            return convertTenUntilTwenty(new String(digits));
+        } else {
+            return convertTens(digits[0])
+                    + (digits[1] != '0' ? "-" + convertUnit(digits[1]) : "");
+        }
+    }
+
+    private String concatWithSpace(String prefix, String postfix) {
+        String wordSeparator = "";
+        if (prefix.length() > 0 && postfix.length() > 0) {
+            wordSeparator = " ";
+        }
+        return prefix + wordSeparator + postfix;
+    }
+
+    private char[] filterInitialZeros(char[] digits) {
+        int cursor = 0;
+        while (cursor < digits.length && digits[cursor] == '0') {
+            cursor++;
+        }
+        boolean almostADigit = cursor < digits.length;
+        if (almostADigit) {
+            return Arrays.copyOfRange(digits, cursor, digits.length);
+        } else {
+            return new char[0];
+        }
+    }
+
+    private char[][] process(char[] digits) {
+        final int digitsLength = digits.length;
+        char[][] result = null;
+        int rest = digitsLength % 3;
+        int quotient = digitsLength / 3;
+        int dimension = quotient + (rest == 0 ? 0 : 1);
+        result = new char[dimension][];
+        result[0] = new char[rest == 0 ? 3 : rest];
+        for (int i = 0; i < dimension - 1; i++) {
+            result[i + 1] = new char[3];
+        }
+        int cursor = digitsLength - 1;
+        int dimensionCursor = dimension - 1;
+        while (dimensionCursor >= 0) {
+            for (int i = result[dimensionCursor].length - 1; i >= 0; i--) {
+                result[dimensionCursor][i] = digits[cursor];
+                cursor--;
+            }
+            dimensionCursor--;
+        }
+        return result;
+    }
+
+    private String order(int length) {
+        switch (length) {
+            case 2:
+                return "thousand";
+            case 3:
+                return "million";
+            case 4:
+                return "billion";
+            default:
+                return "";
+        }
+    }
+
+    private String convertUnit(char digit) {
```

Comment:
If working with longs instead of individual characters, we would be able to use a three digit set (like 123) and then extract the hundreds, tens, and ones digits to correctly say what we want. Something like this:

```java
private String convertNumber(long number) {
  // optionally - throw if number is greater than 999
  int hundreds = number / 100;
  int tens = (number % 100) / 10;
  int ones = number % 10;
  StringBuilder sb = new StringBuilder();
  if (hundreds > 0) {
    sb.append(convertOnes(hundreds)).append(" hundred ")
  }
  if (tens > 0) {
    sb.append(convertTwoDigit(number / 10)).append(" ");
  } else {
    sb.append(convertOnes(ones)).append(" ");
  }
  return sb.toString();
}

String convertTwoDigit(long number) {
  // Optionally throw if number > 99
  // Determine if we are in the 10-19 (which would use a switch-case)
  if (number < 20) return ...
  String multipleOfTen = ... // eg. "forty"
  int ones = number % 10;
  if (ones == 0) {
    return multipleOfTen;
  }
  return multipleOfTen + "-" + convertOnes(ones);
}

String convertOnes(long number) {
  // This one can be a simple switch-case IMO
  ...
}